pkg/clusteragent/autoscaling/workload/pod_watcher.go

```diff
@@ -0,0 +1,163 @@
+	}
+}
+
+// GetPodsForOwner returns the pods for the given owner.
+func (pw *podWatcher) GetPodsForOwner(owner NamespacedPodOwner) []*workloadmeta.KubernetesPod {
+	pw.mutex.RLock()
+	defer pw.mutex.RUnlock()
+	pods, ok := pw.podsPerPodOwner[owner]
+	if !ok {
+		return nil
+	}
+	res := make([]*workloadmeta.KubernetesPod, 0, len(pods))
+	for _, pod := range pods {
+		res = append(res, pod)
+	}
+	return res
+}
+
+// Start subscribes to workloadmeta events and indexes pods by their owner.
+func (pw *podWatcher) Start(ctx context.Context) {
+	log.Debug("Starting PodWatcher")
+	filterParams := workloadmeta.FilterParams{
+		Kinds:     []workloadmeta.Kind{workloadmeta.KindKubernetesPod},
+		Source:    workloadmeta.SourceAll,
+		EventType: workloadmeta.EventTypeAll,
+	}
+	ch := pw.wlm.Subscribe(
+		"app-autoscaler-pod-watcher",
+		workloadmeta.NormalPriority,
+		workloadmeta.NewFilter(&filterParams),
+	)
+	defer pw.wlm.Unsubscribe(ch)
+
+	for {
+		select {
+		case <-ctx.Done():
+			log.Debugf("Stopping PodWatcher")
+			return
+		case eventBundle, more := <-ch:
+			eventBundle.Acknowledge()
+			if !more {
+				log.Debugf("Stopping PodWatcher")
+				return
+			}
+			for _, event := range eventBundle.Events {
+				pw.handleEvent(event)
+			}
+		}
+	}
+}
+
+func (pw *podWatcher) handleEvent(event workloadmeta.Event) {
+	pw.mutex.Lock()
+	switch event.Type {
+	case workloadmeta.EventTypeSet:
+		pw.handleSetEvent(event)
+	case workloadmeta.EventTypeUnset:
+		pw.handleUnsetEvent(event)
+	default:
+		log.Errorf("Ignoring event type %d", event.Type)
+	}
+	pw.mutex.Unlock()
+}
+
+func (pw *podWatcher) handleSetEvent(event workloadmeta.Event) {
+	pod, ok := event.Entity.(*workloadmeta.KubernetesPod)
+	if !ok {
+		log.Debugf("Ignoring event with entity type %T", event.Entity)
+		return
+	}
+	if len(pod.Owners) == 0 {
+		log.Debugf("Ignoring pod %s without owner", pod.Name)
+		return
+	}
+	podOwner := getNamespacedPodOwner(pod.Namespace, &pod.Owners[0])
+	log.Debugf("Adding pod %s to owner %s", pod.ID, podOwner)
+	if _, ok := pw.podsPerPodOwner[podOwner]; !ok {
+		pw.podsPerPodOwner[podOwner] = make(map[string]*workloadmeta.KubernetesPod)
+	}
+	pw.podsPerPodOwner[podOwner][pod.ID] = pod
+}
+
+func (pw *podWatcher) handleUnsetEvent(event workloadmeta.Event) {
+	pod, ok := event.Entity.(*workloadmeta.KubernetesPod)
+	if !ok {
+		log.Debugf("Ignoring event with entity type %T", event.Entity)
+		return
+	}
+	if len(pod.Owners) == 0 {
+		log.Debugf("Ignoring pod %s without owner", pod.Name)
+		return
+	}
+	podOwner := getNamespacedPodOwner(pod.Namespace, &pod.Owners[0])
+	log.Debugf("Removing pod %s from owner %s", pod.ID, podOwner)
+	if _, ok := pw.podsPerPodOwner[podOwner]; !ok {
+		return
+	}
+	delete(pw.podsPerPodOwner[podOwner], pod.ID)
+	if len(pw.podsPerPodOwner[podOwner]) == 0 {
+		delete(pw.podsPerPodOwner, podOwner)
+	}
+}
+
+func getNamespacedPodOwner(ns string, owner *workloadmeta.KubernetesPodOwner) NamespacedPodOwner {
+	res := NamespacedPodOwner{
```

Comment:
Code is slightly incorrect. `ReplicaSet` can be created on their own, we try to guess that in `ParseDeploymentForReplicaSet`, that can return empty.

Reply:
Should we support pods without deployments ?

Reply:
Yes for horizontal we support anything.